File: backend/src/main/java/com/foodrush/service/RestaurantService.java
package com.foodrush.service;

import com.foodrush.common.ApiException;
import com.foodrush.common.Role;
import com.foodrush.dto.Dtos;
import com.foodrush.entity.FoodItemOption;
import com.foodrush.entity.MenuItem;
import com.foodrush.entity.Restaurant;
import com.foodrush.entity.User;
import com.foodrush.repository.MenuItemRepository;
import com.foodrush.repository.RestaurantRepository;
import com.foodrush.repository.ReviewRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.*;

@Service
@RequiredArgsConstructor
public class RestaurantService {

    private final RestaurantRepository restaurantRepository;
    private final MenuItemRepository menuItemRepository;
    private final ReviewRepository reviewRepository;

    public Map<String, Object> list(String q, String cuisine, String city, String open, String sort, int page, int limit) {
        List<Restaurant> items = new ArrayList<>(restaurantRepository.findAll());

        if (q != null && !q.isBlank()) {
            String lower = q.toLowerCase();
            items.removeIf(r -> !r.getName().toLowerCase().contains(lower)
                    && !r.getCuisine().toLowerCase().contains(lower));
        }
        if (cuisine != null && !cuisine.isBlank()) items.removeIf(r -> !r.getCuisine().equals(cuisine));
        if (city != null && !city.isBlank()) items.removeIf(r -> !r.getCity().equalsIgnoreCase(city));
        if (open != null && !open.isBlank()) {
            boolean openFlag = Boolean.parseBoolean(open);
            items.removeIf(r -> r.isOpen() != openFlag);
        }

        String sortKey = sort == null ? "rating" : sort;
        switch (sortKey) {
            case "delivery_time" -> items.sort(Comparator.comparingInt(Restaurant::getDeliveryTimeMinutes));
            case "newest" -> items.sort(Comparator.comparing(Restaurant::getCreatedAt,
                    Comparator.nullsLast(Comparator.reverseOrder())));
            default -> items.sort(Comparator.comparingDouble(Restaurant::getRating).reversed());
        }

        int total = items.size();
        int from = Math.min((page - 1) * limit, total);
        int to = Math.min(from + limit, total);
        List<Restaurant> pageItems = items.subList(from, to);

        return Map.of(
                "items", pageItems,
                "total", total,
                "page", page,
                "limit", limit,
                "pages", (int) Math.ceil((double) total / limit));
    }

    public Map<String, Object> get(Long id) {
        Restaurant r = restaurantRepository.findById(id)
                .orElseThrow(() -> ApiException.notFound("Restaurant not found", "NOT_FOUND"));
        List<MenuItem> menu = menuItemRepository.findByRestaurantIdOrderByCategoryAscNameAsc(id);
        long reviewCount = reviewRepository.countByRestaurantId(id);
        return Map.of("restaurant", r, "menu", menu, "reviewCount", reviewCount);
    }

    public Restaurant create(User actor, Dtos.RestaurantRequest req) {
        if (actor.getRole() != Role.RESTAURANT && actor.getRole() != Role.ADMIN) {
            throw ApiException.forbidden("Only restaurants or admins can create restaurants", "FORBIDDEN");
        }
        Restaurant r = Restaurant.builder()
                .ownerId(actor.getId())
                .name(req.name().trim())
                .description(req.description())
                .cuisine(req.cuisine().trim())
                .imageUrl(req.imageUrl())
                .address(req.address())
                .city(req.city() == null || req.city().isBlank() ? "Lagos" : req.city())
                .phone(req.phone())
                .deliveryTimeMinutes(req.deliveryTimeMinutes() == null ? 30 : req.deliveryTimeMinutes())
                .open(req.isOpen() == null || req.isOpen())
                .approved(actor.getRole() == Role.ADMIN)
                .build();
        return restaurantRepository.save(r);
    }

    @Transactional
    public Restaurant update(User actor, Long id, Dtos.RestaurantRequest req) {
        Restaurant r = restaurantRepository.findById(id)
                .orElseThrow(() -> ApiException.notFound("Restaurant not found", "NOT_FOUND"));
        assertOwner(actor, r);

        if (req.name() != null) r.setName(req.name().trim());
        if (req.description() != null) r.setDescription(req.description());
        if (req.cuisine() != null) r.setCuisine(req.cuisine().trim());
        if (req.imageUrl() != null) r.setImageUrl(req.imageUrl());
        if (req.address() != null) r.setAddress(req.address());
        if (req.city() != null) r.setCity(req.city());
        if (req.phone() != null) r.setPhone(req.phone());
        if (req.deliveryTimeMinutes() != null) r.setDeliveryTimeMinutes(req.deliveryTimeMinutes());
        if (req.isOpen() != null) r.setOpen(req.isOpen());

        return restaurantRepository.save(r);
    }

    @Transactional
    public void approve(User actor, Long id, boolean approved) {
        if (actor.getRole() != Role.ADMIN) {
            throw ApiException.forbidden("Only admins can change approval status", "FORBIDDEN");
        }
        Restaurant r = restaurantRepository.findById(id)
                .orElseThrow(() -> ApiException.notFound("Restaurant not found", "NOT_FOUND"));
        r.setApproved(approved);
        restaurantRepository.save(r);
    }

    public void delete(User actor, Long id) {
        Restaurant r = restaurantRepository.findById(id)
                .orElseThrow(() -> ApiException.notFound("Restaurant not found", "NOT_FOUND"));
        assertOwner(actor, r);
        restaurantRepository.delete(r);
    }

    // ---- Menu items ----

    @Transactional
    public MenuItem addMenuItem(User actor, Long restaurantId, Dtos.MenuItemRequest req) {
        Restaurant r = restaurantRepository.findById(restaurantId)
                .orElseThrow(() -> ApiException.notFound("Restaurant not found", "NOT_FOUND"));
        assertOwner(actor, r);

        MenuItem item = MenuItem.builder()
                .restaurant(r)
                .name(req.name().trim())
                .description(req.description())
                .price(req.price())
                .category(req.category())
                .imageUrl(req.imageUrl())
                .available(req.isAvailable() == null || req.isAvailable())
                .build();

        if (req.options() != null) {
            for (Dtos.OptionRequest o : req.options()) {
                item.getOptions().add(FoodItemOption.builder()
                        .menuItem(item).name(o.name().trim()).price(o.price()).build());
            }
        }
        return menuItemRepository.save(item);
    }

    @Transactional
    public MenuItem updateMenuItem(User actor, Long itemId, Dtos.MenuItemRequest req) {
        MenuItem item = menuItemRepository.findById(itemId)
                .orElseThrow(() -> ApiException.notFound("Menu item not found", "NOT_FOUND"));
        assertOwner(actor, item.getRestaurant());

        if (req.name() != null) item.setName(req.name().trim());
        if (req.description() != null) item.setDescription(req.description());
        if (req.price() != 0) item.setPrice(req.price());
        if (req.category() != null) item.setCategory(req.category());
        if (req.imageUrl() != null) item.setImageUrl(req.imageUrl());
        if (req.isAvailable() != null) item.setAvailable(req.isAvailable());

        if (req.options() != null) {
            item.getOptions().clear();
            for (Dtos.OptionRequest o : req.options()) {
                item.getOptions().add(FoodItemOption.builder()
                        .menuItem(item).name(o.name().trim()).price(o.price()).build());
            }
        }
        return menuItemRepository.save(item);
    }

    public void deleteMenuItem(User actor, Long itemId) {
        MenuItem item = menuItemRepository.findById(itemId)
                .orElseThrow(() -> ApiException.notFound("Menu item not found", "NOT_FOUND"));
        assertOwner(actor, item.getRestaurant());
        menuItemRepository.delete(item);
    }

    public Map<String, Object> menu(Long id) {
        restaurantRepository.findById(id)
                .orElseThrow(() -> ApiException.notFound("Restaurant not found", "NOT_FOUND"));
        return Map.of("restaurantId", id, "items",
                menuItemRepository.findByRestaurantIdOrderByCategoryAscNameAsc(id));
    }

    public Map<String, Object> categories() {
        Map<String, Long> counts = new HashMap<>();
        for (MenuItem m : menuItemRepository.findAll()) {
            if (m.getCategory() != null && !m.getCategory().isBlank()) {
                counts.merge(m.getCategory(), 1L, Long::sum);
            }
        }
        List<Map<String, Object>> list = new ArrayList<>();
        counts.forEach((name, n) -> list.add(Map.of("name", name, "itemCount", n)));
        list.sort((a, b) -> Long.compare((Long) b.get("itemCount"), (Long) a.get("itemCount")));
        return Map.of("categories", list);
    }

    private void assertOwner(User actor, Restaurant r) {
        if (actor.getRole() == Role.ADMIN) return;
        if (r.getOwnerId() == null || !r.getOwnerId().equals(actor.getId())) {
            throw ApiException.forbidden("You do not own this restaurant", "FORBIDDEN");
        }
    }
}

File: src/services/restaurant.service.ts
import {
  menuRepository,
  restaurantRepository,
} from "@/repositories";
import type {
  MenuItemEntity,
  RestaurantEntity,
  RestaurantRepository,
  MenuRepository,
} from "@/repositories/contracts";
import { query, queryValue } from "@/db/raw";
import { ApiError } from "@/lib/http";
import type { UserEntity } from "@/repositories/contracts";

export class RestaurantService {
  constructor(
    private readonly restaurants: RestaurantRepository = restaurantRepository,
    private readonly menuRepo: MenuRepository = menuRepository,
  ) {}

  async list(filter: {
    q?: string;
    cuisine?: string;
    city?: string;
    open?: boolean;
    sort?: string;
    page: number;
    limit: number;
  }) {
    return this.restaurants.list({
      q: filter.q,
      cuisine: filter.cuisine,
      city: filter.city,
      open: filter.open,
      sort: filter.sort ?? "rating",
      page: filter.page,
      limit: filter.limit,
    });
  }

  async get(id: number) {
    const restaurant = await this.restaurants.findById(id);
    if (!restaurant) throw new ApiError(404, "Restaurant not found", "NOT_FOUND");

    const items = await this.menuRepo.findByRestaurant(id);
    const reviewCount = (await queryValue<number>(`SELECT count(*) FROM reviews WHERE restaurant_id = $1`, [id])) ?? 0;

    return {
      restaurant,
      menu: await this.attachOptions(items),
      reviewCount,
    };
  }

  async menu(id: number) {
    const restaurant = await this.restaurants.findById(id);
    if (!restaurant) throw new ApiError(404, "Restaurant not found", "NOT_FOUND");
    const items = await this.menuRepo.findByRestaurant(id);
    return { restaurantId: id, items: await this.attachOptions(items) };
  }

  async create(actor: UserEntity, input: {
    name: string;
    description?: string;
    cuisine: string;
    imageUrl?: string;
    coverImageUrl?: string;
    address: string;
    city?: string;
    latitude?: number;
    longitude?: number;
    phone?: string;
    deliveryTimeMinutes?: number;
    minimumOrder?: number;
    isOpen?: boolean;
  }): Promise<RestaurantEntity> {
    if (actor.role !== "restaurant" && actor.role !== "admin") {
      throw new ApiError(403, "Only restaurants or admins can create restaurants", "FORBIDDEN");
    }

    return this.restaurants.create({
      ownerId: actor.id,
      name: input.name.trim(),
      description: input.description,
      cuisine: input.cuisine.trim(),
      imageUrl: input.imageUrl,
      coverImageUrl: input.coverImageUrl,
      address: input.address,
      city: input.city,
      latitude: input.latitude,
      longitude: input.longitude,
      phone: input.phone,
      deliveryTimeMinutes: input.deliveryTimeMinutes,
      minimumOrder: input.minimumOrder,
      isOpen: input.isOpen,
      isApproved: actor.role === "admin",
    });
  }

  async update(actor: UserEntity, id: number, patch: Partial<RestaurantEntity>): Promise<RestaurantEntity> {
    const restaurant = await this.restaurants.findById(id);
    if (!restaurant) throw new ApiError(404, "Restaurant not found", "NOT_FOUND");
    this.assertOwner(actor, restaurant);

    if (patch.isApproved !== undefined && actor.role !== "admin") {
      throw new ApiError(403, "Only admins can change approval status", "FORBIDDEN");
    }

    const updated = await this.restaurants.update(id, patch);
    if (!updated) throw new ApiError(404, "Restaurant not found", "NOT_FOUND");
    return updated;
  }

  async remove(actor: UserEntity, id: number): Promise<void> {
    const restaurant = await this.restaurants.findById(id);
    if (!restaurant) throw new ApiError(404, "Restaurant not found", "NOT_FOUND");
    this.assertOwner(actor, restaurant);
    await this.restaurants.remove(id);
  }

  async addMenuItem(actor: UserEntity, restaurantId: number, input: {
    name: string;
    description?: string;
    price: number;
    category?: string;
    imageUrl?: string;
    isAvailable?: boolean;
    options?: { name: string; price: number }[];
  }): Promise<MenuItemEntity> {
    const restaurant = await this.restaurants.findById(restaurantId);
    if (!restaurant) throw new ApiError(404, "Restaurant not found", "NOT_FOUND");
    this.assertOwner(actor, restaurant);

    const item = await this.menuRepo.create({
      restaurantId,
      name: input.name.trim(),
      description: input.description,
      price: input.price,
      category: input.category,
      imageUrl: input.imageUrl,
      isAvailable: input.isAvailable,
    });

    if (input.options && input.options.length > 0) {
      for (const o of input.options) {
        await query(`INSERT INTO food_item_options (menu_item_id, name, price) VALUES ($1, $2, $3)`, [item.id, o.name, o.price]);
      }
    }
    return item;
  }

  async categories() {
    const rows = await query<{ category: string; n: number }>(
      `SELECT category, count(*) AS n FROM menu_items WHERE is_available = true AND category IS NOT NULL GROUP BY category ORDER BY n DESC`,
    );
    return rows.map((r) => ({ name: r.category, itemCount: Number(r.n) }));
  }

  private async attachOptions(items: MenuItemEntity[]) {
    const ids = items.map((i) => i.id);
    const options = await this.menuRepo.optionsForItems(ids);
    const byItem = new Map<number, typeof options>();
    for (const o of options) {
      const arr = byItem.get(o.menuItemId) ?? [];
      arr.push(o);
      byItem.set(o.menuItemId, arr);
    }
    return items.map((item) => ({ ...item, options: byItem.get(item.id) ?? [] }));
  }

  private assertOwner(actor: UserEntity, restaurant: RestaurantEntity): void {
    if (actor.role === "admin") return;
    if (restaurant.ownerId !== actor.id) {
      throw new ApiError(403, "You do not own this restaurant", "FORBIDDEN");
    }
  }
}

export const restaurantService = new RestaurantService();
